build error: SWIG_Python_str_AsChar was not declared in this scope

Fedora f40 has upgraded a few packages in the last week and I'm seeing this error at the end of the build. These are some changes since the last f40 build:

* gcc-13.2.1 > gcc-14.0.1
* boost-1.81.0 > boost-1.83.0
* swig-4.1.1 > swig-4.2.0
* python-3.12 > python-3.12.1

```
[100%] Building CXX object ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx.o
cd /builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap && /usr/bin/g++ -DBOOST_ALL_NO_LIB -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAS_SCHEMA_2x3 -DHAS_SCHEMA_4 -DHAS_SCHEMA_4x1 -DHAS_SCHEMA_4x2 -DHAS_SCHEMA_4x3 -DHAS_SCHEMA_4x3_add1 -DHAS_SCHEMA_4x3_add2 -DHAS_SCHEMA_4x3_rc1 -DHAS_SCHEMA_4x3_rc2 -DHAS_SCHEMA_4x3_rc3 -DHAS_SCHEMA_4x3_rc4 -DHAS_SCHEMA_4x3_tc1 -DIFC_SHARED_BUILD -DSCHEMA_SEQ="(2x3)(4)(4x1)(4x2)(4x3_rc1)(4x3_rc2)(4x3_rc3)(4x3_rc4)(4x3)(4x3_tc1)(4x3_add1)(4x3_add2)" -DUSE_MMAP -DWITH_GLTF -DWITH_HDF5 -DWITH_IFCXML -D_ifcopenshell_wrapper_EXPORTS -I/usr/include/opencascade -I/usr/include/libxml2 -I/usr/include/python3.12 -I/builddir/build/BUILD/IfcOpenShell-0.7.0/src/ifcwrap -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64   -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -DNDEBUG -std=gnu++14 -fPIC   -Wall -Wextra -Wno-maybe-uninitialized -Wno-deprecated-copy -fPIC -MD -MT ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx.o -MF CMakeFiles/_ifcopenshell_wrapper.dir/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx.o.d -o CMakeFiles/_ifcopenshell_wrapper.dir/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx.o -c /builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx: In function â€˜T cast_pyobject(PyObject*) [with T = std::__cxx11::basic_string<char>; PyObject = _object]â€™:
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:4138: error: â€˜SWIG_Python_str_AsCharâ€™ was not declared in this scope; did you mean â€˜SWIG_Python_str_FromCharâ€™?
 4138 |                 char* str_data = SWIG_Python_str_AsChar(element);
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:4140: error: â€˜SWIG_Python_str_DelForPy3â€™ was not declared in this scope; did you mean â€˜SWIG_Python_str_FromFormatâ€™?
 4140 |                 SWIG_Python_str_DelForPy3(str_data);
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx: At global scope:
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:7323: warning: type qualifiers ignored on function return type [-Wignored-qualifiers]
 7323 | SWIGINTERN char const *const IfcUtil_IfcBaseClass_get_argument_type(IfcUtil::IfcBaseClass const *self,unsigned int i){
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx: In function â€˜std::vector<std::__cxx11::basic_string<char> > IfcParse_enumeration_type_argument_types(IfcParse::enumeration_type*)â€™:
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:7537: warning: unused parameter â€˜selfâ€™ [-Wunused-parameter]
 7537 | SWIGINTERN std::vector< std::string,std::allocator< std::string > > IfcParse_enumeration_type_argument_types(IfcParse::enumeration_type *self){
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx: In function â€˜PyObject* convert_cpp_attribute_to_python(IfcUtil::ArgumentType, Argument&)â€™:
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:7640: warning: enumeration value â€˜Argument_NULLâ€™ not handled in switch [-Wswitch]
 7640 |                 switch(type) {
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:7640: warning: enumeration value â€˜Argument_DERIVEDâ€™ not handled in switch [-Wswitch]
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:7640: warning: enumeration value â€˜Argument_AGGREGATE_OF_EMPTY_AGGREGATEâ€™ not handled in switch [-Wswitch]
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:7640: warning: enumeration value â€˜Argument_UNKNOWNâ€™ not handled in switch [-Wswitch]
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx: In function â€˜PyObject* _wrap_delete_IfcBaseInterface(PyObject*, PyObject*)â€™:
/builddir/build/BUILD/IfcOpenShell-0.7.0/cmake/redhat-linux-build/ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx:53675: warning: deleting object of abstract class type â€˜IfcUtil::IfcBaseInterfaceâ€™ which has non-virtual destructor will cause undefined behavior [-Wdelete-non-virtual-dtor]
```

--- a/src/ifcwrap/utils/type_conversion.i
+++ b/src/ifcwrap/utils/type_conversion.i
@@ -72,9 +72,16 @@
 
 	template <>
 	std::string cast_pyobject(PyObject* element) {
+	#if SWIG_VERSION >= 0x040200
+	    PyObject *pbytes = NULL;
+	    const char* str_data = SWIG_PyUnicode_AsUTF8AndSize(element, NULL, &pbytes);
+		std::string str = str_data;
+		Py_XDECREF(pbytes);
+	#else
 		char* str_data = SWIG_Python_str_AsChar(element);
 		std::string str = str_data;
 		SWIG_Python_str_DelForPy3(str_data);
+	#endif
 		return str;
 	}
 

